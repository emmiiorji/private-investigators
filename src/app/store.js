import { configureStore } from '@reduxjs/toolkit';
import investigatorsReducer from './redux/investigator/investigatorSlice';

const store = configureStore({
  reducer: {
    investigators: investigatorsReducer,
  },
});

export default store;
