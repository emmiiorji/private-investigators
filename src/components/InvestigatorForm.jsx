import React from 'react';

const InvestigatorForm = () => (<h1>Investigator Form</h1>);

export default InvestigatorForm;
