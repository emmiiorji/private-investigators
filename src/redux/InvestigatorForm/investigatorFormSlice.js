/* eslint-disable no-param-reassign */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const createInvestigator = createAsyncThunk('investigators/create', async (investigator) => {
  const response = await axios.post(`${process.env.REACT_APP_SERVER_BASE_URL}/v1/investigators/`, {
    investigator: {
      name: investigator.name,
      photo: investigator.photo,
      description: investigator.description,
      fee: investigator.fee,
      rating: investigator.rating,
    },
  }, { withCredentials: true });
  return response.data;
});

const initialState = {
  status: 'idle',
  value: [],
};

const createInvestigatorsSlice = createSlice({
  name: 'investigators',
  initialState,
  reducers: {
    loadInvestigators: (state, action) => {
      const newState = { ...state, value: [...state.value, action.payload] };
      return newState;
    },
  },

  extraReducers: (builder) => {
    builder
      .addCase(createInvestigator.fulfilled, (state, action) => {
        const newState = { ...state, status: 'succeeded', value: action.payload };
        return newState;
      });
  },
});

export default createInvestigatorsSlice.reducer;
