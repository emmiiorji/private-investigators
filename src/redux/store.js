import { configureStore } from '@reduxjs/toolkit';
// import userReducer from './CurrentUser/CurrentUserSlice';
import investigatorsReducer from './investigator/investigatorSlice';
import appointmentsReducer from './Appointments/appointmentsSlice';
import userReducer from './CurrentUser/CurrentUserSlice';

const store = configureStore({
  reducer: {
    investigators: investigatorsReducer,
    appointments: appointmentsReducer,
    userReducer,
  },
});

export default store;
