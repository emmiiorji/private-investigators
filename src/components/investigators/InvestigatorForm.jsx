import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import Swal from 'sweetalert2';
import { Form, Button } from 'react-bootstrap';
import { createInvestigator } from '../../redux/InvestigatorForm/investigatorFormSlice';
import { fetchInvestigators } from '../../redux/investigators/investigatorSlice';

const InvestigatorForm = () => {
  const [state, setState] = useState({
    name: '',
    photo: '',
    description: '',
    fee: 0,
    rating: 0,
  });
  const cInvestigator = useSelector((state) => state.createInvestigatorsReducer);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (cInvestigator.status === 'succeeded') {
      Swal.fire({
        position: 'top',
        icon: 'success',
        title: 'Investigator created successfully!',
        showConfirmButton: false,
        timer: 2000,
      });
      dispatch(fetchInvestigators());
      navigate('/app/investigators');
    }
  }, [cInvestigator, navigate, dispatch]);

  const handleChange = (event, name) => {
    if (name === 'name' || name === 'photo' || name === 'description' || name === 'fee' || name === 'rating') {
      setState((state) => ({
        ...state,
        [name]: event.target.value,
      }));
    }
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (state.name && state.photo && state.description && state.fee && state.rating) {
      dispatch(createInvestigator({ ...state }));
    }
  };

  return (
    <div
      className=""
      style={{
        backgroundImage: 'url("https://mdbootstrap.com/img/Photos/Others/images/76.jpg")',
        height: '100vh',
        width: '100%',
      }}
    >
      <div
        className="d-flex flex-column justify-content-center align-items-center opacity-75"
        style={{
          height: '100vh',
          backgroundColor: '#97BF0F',
          color: 'white',
        }}
      >
        <h1>Add an Investigator</h1>
        <hr className="border border-2 w-25" />
        <p className="mb-5">Use the following form to add a new investigator.</p>
        <Form onSubmit={handleSubmit} className="row gap-3">
          <Form.Control className="col rounded-pill border border-white text-white py-3 px-3" style={{ backgroundColor: '#97BF0F' }} placeholder="Name" type="text" value={state.name} required onChange={(e) => handleChange(e, 'name')} />
          <Form.Control className="col rounded-pill border border-white text-white py-3 px-3" style={{ backgroundColor: '#97BF0F' }} placeholder="Photo link" type="text" value={state.photo} required onChange={(e) => handleChange(e, 'photo')} />
          <Form.Control className="col rounded-pill border border-white text-white py-3 px-3" style={{ backgroundColor: '#97BF0F' }} placeholder="Description" type="text" value={state.description} required onChange={(e) => handleChange(e, 'description')} />
          <Form.Control className="col rounded-pill border border-white text-white py-3 px-3" style={{ backgroundColor: '#97BF0F' }} placeholder="Fee" type="number" value={state.fee} required onChange={(e) => handleChange(e, 'fee')} />
          <Form.Control className="col rounded-pill border border-white text-white py-3 px-3" style={{ backgroundColor: '#97BF0F' }} placeholder="Rating" type="number" value={state.rating} required onChange={(e) => handleChange(e, 'rating')} />
          <Button type="submit" className="rounded-pill border border-white" style={{ backgroundColor: 'white', color: '#97BF0F', fontWeight: 'bold' }}>Submit</Button>
        </Form>
      </div>
    </div>
  );
};

export default InvestigatorForm;
