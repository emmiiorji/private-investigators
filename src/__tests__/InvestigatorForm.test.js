import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';
import InvestigatorForm from '../components/investigators/InvestigatorForm';

describe('Test adding new investigator', () => {
  let addInvestigator;

  beforeAll(async () => {
    addInvestigator = renderer
      .create(
        <Provider store={store}>
          <Router>
            <InvestigatorForm />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  it('It should render correctly', () => {
    expect(addInvestigator).toBeTruthy();
  });

  it('Should match its snapshot', () => {
    expect(addInvestigator).toMatchSnapshot();
  });

  it('It renders the form', () => {
    // to add the data-testid attribute for the form
    const form = screen.findByTestId('add-form');
    waitFor(() => expect(form).toBeInTheDocument());
  });

  it('Should have a button with Add text', () => {
    // to add the data-testid attribute for the button
    const btn = screen.findByTestId('add-btn');
    waitFor(() => expect(btn).toHaveTextContent('Add'));
  });
});
