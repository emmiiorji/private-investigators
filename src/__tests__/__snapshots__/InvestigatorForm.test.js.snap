// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test adding new investigator Should match its snapshot 1`] = `
<h1>
  Investigator Form
</h1>
`;
